PWGJE/FlavourJetTasks/AliAnalysisTaskHFJetIPQA.cxx

```diff
@@ -1654,166 +1734,120 @@ Bool_t AliAnalysisTaskHFJetIPQA::Run(){
-
-                //_____________________________
-                //V0 tag decisions
-                //printf("isV0=%i, jetflaovur=%i, jetpt=%f", );
-                bool isV0Jet=kFALSE;
-                if((hasIPs[0])&&(!fIsPythia)&&(sImpParXYSig[0].is_V0==V0Rec)&&(fApplyV0Rej==V0JetRej)){
-                    isV0Jet=kTRUE;
-                }
-                //printf("New jetflavour=%i, isV0Jet=%i\n",fJetFlavour, isV0Jet);
+        //FillHist("fh1dParticlesPerJet",NJetParticles,1);
 
-                if(fIsPythia){
-                  if(hasIPs[0])FillV0EfficiencyHists(sImpParXYSig[0].is_V0, fJetFlavour, fJetRecPt, isV0Jet);
-                }
-                //_________________________________________
-                //TAGGING
-                if(fDoTCTagging!=TCNo&&fDoProbTagging!=ProbNo) AliFatal("Don't do track counting and probability tagging simultaneously!");
-
-                bool ** kTagDec=new bool*[fNThresholds];
-                for(int iThresh=0;iThresh<fNThresholds;iThresh++){
-                  kTagDec[iThresh]=new bool[6];
-                  for(int iType=0;iType<6;iType++){
-                    kTagDec[iThresh][iType]=0;
-                  }
-                }
-                //**************
-                //MC Track Counting
-                if(fDoTCTagging!=TCNo){
-                  if(fIsPythia||((!fIsPythia)&&(!isV0Jet))){
-                      //printf("isV0Jet=%i\n", isV0Jet);
-                      if(fUseSignificance){DoTCTagging(fJetRecPt, hasIPs,ipvalsig, kTagDec);}
-                      else{DoTCTagging(fJetRecPt, hasIPs,ipval, kTagDec);}
-                  }
-                }
-                //**************
-                //Probability Dists
-                fJetProb=GetTrackProbability(fJetRecPt,hasIPs, ipvalsig);
-                //Generation of Track Probability Hists
-
-                tJetTree->Fill();
-                sImpParXY.clear();
-                sImpParXYSig.clear();
-                sImpParXYZ.clear();
-                sImpParXYZSig.clear();
-
-                for(int iThresh=0;iThresh<fNThresholds;iThresh++){
-                  delete kTagDec[iThresh];
-                }
-                delete [] kTagDec;
-              }//end jetloop*/
-            return kTRUE;
+        DetermineIPVars(sImpParXY, sImpParXYSig, ipvalsig, ipval, nGoodIPTracks);
+
+        //_____________________________
+        //V0 tag decisions
+        //printf("isV0=%i, jetflaovur=%i, jetpt=%f", );
+        if((nGoodIPTracks>0)&&(!fIsPythia)&&(sImpParXYSig[0].is_V0==V0Rec)&&(fApplyV0Rej==V0JetRej)){
+          isV0Jet=kTRUE;
+        }
+        //printf("New jetflavour=%i, isV0Jet=%i\n",fJetFlavour, isV0Jet);
+        if(fIsPythia){
+          if(nGoodIPTracks>0)FillV0EfficiencyHists(sImpParXYSig[0].is_V0, fJetFlavour, fJetRecPt, isV0Jet);
+        }
+
+        //_________________________________________
+        //TAGGING
+        Bool_t ** kTagDec=new Bool_t*[fNThresholds];
```

Comment:
If you use std::vector<std::vector<bool>> for your 2D dynamic array you could get rid of the manual memory management. Vectors can be initialized with fixed initial sizes.